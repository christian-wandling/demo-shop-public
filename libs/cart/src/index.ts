export * from './lib/cart.routes';
export * from './lib/cart.facade';
export * from './lib/components/cart/cart.component';
export * from './lib/components/checkout/checkout.component';
export * from './lib/components/cart-icon/cart-icon.component';
