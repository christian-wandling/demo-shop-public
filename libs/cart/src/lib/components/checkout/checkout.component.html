<div class="bg-gray-100 sm:m-2 lg:m-6 p-4 sm:p-8 lg:p-10 sm:rounded">
  <div class="flex flex-col lg:flex-row justify-between gap-8">
    <form [formGroup]="checkoutForm()">
      <div class="space-y-8">
        <div>
          <div
            class="flex justify-between items-center p-4 bg-gray-100 cursor-pointer lg:pointer-events-none"
            (click)="shippingInformationExtended.set(!shippingInformationExtended())">
            <h2 class="text-base font-semibold leading-7 text-gray-900">Shipping information</h2>
            <svg
              xmlns="http://www.w3.org/2000/svg"
              class="lg:hidden w-5 h-5 transition-transform duration-300"
              [ngClass]="{ 'rotate-180': shippingInformationExtended() }"
              fill="none"
              viewBox="0 0 24 24"
              stroke="currentColor">
              <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M19 9l-7 7-7-7" />
            </svg>
          </div>
          <div
            class="mt-4 lg:max-h-[1000px] grid md:grid-cols-8 gap-x-6 gap-y-4 transition-all duration-500 ease-in-out overflow-hidden"
            [ngClass]="{
              'max-h-0': !shippingInformationExtended(),
              'max-h-[1000px]': shippingInformationExtended()
            }">
            <div class="md:col-span-4">
              <label for="first-name" class="required block text-sm font-medium leading-6 text-gray-500 disabled">
                First name
              </label>
              <div class="mt-2">
                <input
                  formControlName="firstname"
                  type="text"
                  name="first-name"
                  id="first-name"
                  autocomplete="given-name"
                  class="block w-full rounded-md border-0 py-1.5 text-gray-500 bg-gray-200 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6"
                  data-testid="checkout-form-element-firstname" />
              </div>
              <lib-form-error [errors]="checkoutForm().controls.firstname.errors" />
            </div>

            <div class="md:col-span-4">
              <label for="last-name" class="required block text-sm font-medium leading-6 text-gray-500">
                Last name
              </label>
              <div class="mt-2">
                <input
                  formControlName="lastname"
                  type="text"
                  name="last-name"
                  id="last-name"
                  autocomplete="family-name"
                  class="block w-full rounded-md border-0 py-1.5 text-gray-500 bg-gray-200 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6"
                  data-testid="checkout-form-element-lastname" />
                <lib-form-error [errors]="checkoutForm().controls.firstname.errors" />
              </div>
            </div>

            <div class="md:col-span-4">
              <label for="email" class="required block text-sm font-medium leading-6 text-gray-500">
                Email address
              </label>
              <div class="mt-2">
                <input
                  formControlName="email"
                  id="email"
                  name="email"
                  type="email"
                  autocomplete="email"
                  class="block w-full rounded-md border-0 py-1.5 text-gray-500 bg-gray-200 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6"
                  data-testid="checkout-form-element-email" />
              </div>
              <lib-form-error [errors]="checkoutForm().controls.email.errors" />
            </div>

            <div class="md:col-span-4">
              <label for="phone" class="block text-sm font-medium leading-6 text-gray-900">Phone</label>
              <div class="mt-2">
                <input
                  formControlName="phone"
                  id="phone"
                  name="phone"
                  autocomplete="tel"
                  class="block w-full rounded-md border-0 py-1.5 text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6"
                  data-testid="checkout-form-element-phone" />
              </div>
            </div>

            <ng-container [formGroup]="checkoutForm().controls.address">
              <div class="md:col-span-4">
                <label for="country" class="required block text-sm font-medium leading-6 text-gray-900">Country</label>
                <div class="mt-2">
                  <select
                    formControlName="country"
                    id="country"
                    name="country"
                    autocomplete="country-name"
                    class="block w-full rounded-md border-0 py-1.5 text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:max-w-xs sm:text-sm sm:leading-6"
                    data-testid="checkout-form-element-country">
                    <option>United States</option>
                    <option>Austria</option>
                    <option>Vietnam</option>
                  </select>
                </div>
                <lib-form-error [errors]="checkoutForm().controls.address.controls.country.errors" />
              </div>

              <div class="md:col-span-5">
                <label for="street" class="required block text-sm font-medium leading-6 text-gray-900">Street</label>
                <div class="mt-2">
                  <input
                    formControlName="street"
                    type="text"
                    name="street"
                    id="street"
                    autocomplete="address-line1"
                    class="block w-full rounded-md border-0 py-1.5 text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6"
                    data-testid="checkout-form-element-street" />
                </div>
              </div>

              <div class="md:col-span-3">
                <label for="apartment" class="required block text-sm font-medium leading-6 text-gray-900">
                  Apartment
                </label>
                <div class="mt-2">
                  <input
                    formControlName="apartment"
                    type="text"
                    name="apartment"
                    id="apartment"
                    autocomplete="address-line2"
                    class="block w-full rounded-md border-0 py-1.5 text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6"
                    data-testid="checkout-form-element-apartment" />
                </div>
                <lib-form-error [errors]="checkoutForm().controls.address.controls.apartment.errors" />
              </div>

              <div class="md:col-span-3">
                <label for="city" class="required block text-sm font-medium leading-6 text-gray-900">City</label>
                <div class="mt-2">
                  <input
                    formControlName="city"
                    type="text"
                    name="city"
                    id="city"
                    autocomplete="address-level2"
                    class="block w-full rounded-md border-0 py-1.5 text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6"
                    data-testid="checkout-form-element-city" />
                </div>
                <lib-form-error [errors]="checkoutForm().controls.address.controls.city.errors" />
              </div>

              <div class="md:col-span-3">
                <label for="region" class="block text-sm font-medium leading-6 text-gray-900">State / Province</label>
                <div class="mt-2">
                  <input
                    formControlName="region"
                    type="text"
                    name="region"
                    id="region"
                    autocomplete="address-level1"
                    class="block w-full rounded-md border-0 py-1.5 text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6"
                    data-testid="checkout-form-element-region" />
                </div>
              </div>

              <div class="md:col-span-2">
                <label for="zip" class="required block text-sm font-medium leading-6 text-gray-900">
                  ZIP / Postal code
                </label>
                <div class="mt-2">
                  <input
                    formControlName="zip"
                    type="text"
                    name="zip"
                    id="zip"
                    autocomplete="postal-code"
                    class="block w-full rounded-md border-0 py-1.5 text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6"
                    data-testid="checkout-form-element-zip" />
                </div>
                <lib-form-error [errors]="checkoutForm().controls.address.controls.zip.errors" />
              </div>
            </ng-container>

            <button
              (click)="updateUser()"
              [disabled]="!updateButtonEnabled()"
              [ngClass]="{
                'bg-gray-200 text-gray-800 pointer-events-none': !updateButtonEnabled(),
                'bg-indigo-600 text-white hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-indigo-500 focus:ring-offset-2':
                  updateButtonEnabled()
              }"
              class="btn-update md:col-span-full float-right m-2 mt-6 lg:max-w-60 flex flex-1 items-center justify-center rounded-md border border-transparent px-8 py-3 text-base font-medium"
              data-testid="checkout-update-shipping-information-button">
              Update
            </button>
          </div>
        </div>
      </div>
    </form>

    <section>
      <h2 class="text-lg font-medium text-gray-900" id="slide-over-title">Order Summary</h2>
      <lib-cart-items [items]="items()" (removeItem)="removeItem($event)" />
      <div class="border-t border-gray-900/10 mt-12">
        <div class="mt-8 flex justify-between">
          <span>Total</span><span data-testid="checkout-total-price">{{ price() | currency }}</span>
        </div>
      </div>

      <section>
        <div class="mt-6 flex items-center justify-center gap-x-6">
          <button
            (click)="checkout()"
            [disabled]="!checkoutButtonEnabled()"
            [ngClass]="{
              'bg-gray-200 text-gray-800 pointer-events-none': !checkoutButtonEnabled,
              'bg-indigo-600 text-white hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-indigo-500 focus:ring-offset-2':
                checkoutButtonEnabled()
            }"
            class="btn-checkout mt-6 lg:max-w-60 flex w-full items-center justify-center rounded-md border border-transparent px-8 py-3 text-base font-medium"
            data-testid="checkout-finish-button">
            Checkout
          </button>
        </div>
      </section>
    </section>
  </div>
</div>
