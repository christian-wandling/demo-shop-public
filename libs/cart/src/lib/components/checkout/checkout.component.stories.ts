import { applicationConfig, Meta, moduleMetadata, StoryObj } from '@storybook/angular';
import { CommonModule, NgOptimizedImage } from '@angular/common';
import { provideRouter } from '@angular/router';
import { signal } from '@angular/core';
import { AddressResponse, CartItemResponse, UserResponse } from '@demo-shop/api';
import { faker } from '@faker-js/faker';
import { FormErrorComponent, provideImageLoader } from '@demo-shop/shared';
import { CartFacade } from '../../cart.facade';
import { CheckoutComponent } from '../../cart.routes';
import { CartItemsComponent } from '../shared/cart-items/cart-items.component';
import { within } from '@storybook/testing-library';
import { expect } from '@storybook/jest';
import { ReactiveFormsModule } from '@angular/forms';
import { UserFacade } from '@demo-shop/user';

const items: CartItemResponse[] = Array.from({ length: 5 }).map((_, id) => {
  const quantity = faker.number.int({ min: 1, max: 5 });
  const unitPrice = faker.number.int({ min: 1, max: 20 }) * 10 - 1;

  return {
    id,
    productId: faker.number.int({ min: 1, max: 20 }),
    productName: faker.commerce.productName(),
    productThumbnail: faker.image.urlPicsumPhotos(),
    quantity,
    unitPrice,
    totalPrice: quantity * unitPrice,
  };
});

const address: AddressResponse = {
  street: faker.location.streetAddress(),
  apartment: faker.number.int({ max: 100 }).toString(),
  city: faker.location.city(),
  zip: faker.location.zipCode('#####'),
  region: faker.location.state(),
  country: 'United States',
};

const user: UserResponse = {
  id: 1,
  email: faker.internet.email(),
  firstname: faker.person.firstName(),
  lastname: faker.person.lastName(),
  phone: faker.helpers.fromRegExp('+[0-9]{9}'),
};

const mockCartFacade = (items: CartItemResponse[]) => ({
  removeItem: () => {
    return;
  },
  checkout: () => {
    return;
  },
  getAll: () => signal(items),
  getTotalPrice: () => {
    const totalPrice = items.reduce((acc, curr) => acc + curr.totalPrice, 0);
    return signal(totalPrice);
  },
  getShowCart: () => signal(true),
  setShowCart: () => {
    return;
  },
});

const mockUserFacade = (user: UserResponse) => ({
  updateUserAddress: () => {
    return;
  },
  updateUserPhone: () => {
    return;
  },
  getCurrentUser: () => signal(user),
});

const meta: Meta<CheckoutComponent> = {
  component: CheckoutComponent,
  title: 'Cart/CheckoutComponent',
  decorators: [
    applicationConfig({
      providers: [provideImageLoader(), provideRouter([]), { provide: CartFacade, useValue: mockCartFacade(items) }],
    }),
    moduleMetadata({
      imports: [CommonModule, NgOptimizedImage, ReactiveFormsModule, CartItemsComponent, FormErrorComponent],
    }),
  ],
  argTypes: {
    removeItem: { action: 'removeItem', control: false },
    checkout: { action: 'checkout', control: false },
    updateUser: { action: 'updateUser', control: false },
  },
  parameters: {
    controls: {
      exclude: [
        'checkoutForm',
        'shippingInformationExtended',
        '#cartFacade',
        'items',
        'price',
        '#userFacade',
        'user',
        '#fb',
        '#router',
        'formDirty',
        'formValid',
        'updateButtonEnabled',
        'checkoutButtonEnabled',
        'createCheckOutForm',
      ],
    },
  },
};
export default meta;
type Story = StoryObj<CheckoutComponent>;

export const WithValidShippingInformation: Story = {
  decorators: [
    applicationConfig({
      providers: [{ provide: UserFacade, useValue: mockUserFacade({ ...user, address }) }],
    }),
  ],
  play: async ({ canvasElement, args }) => {
    const canvas = within(canvasElement);
    const firstname: HTMLInputElement = canvas.getByRole('textbox', { name: /first name/i });
    const lastname: HTMLInputElement = canvas.getByRole('textbox', { name: /last name/i });
    const email: HTMLInputElement = canvas.getByRole('textbox', { name: /email/i });
    const phone: HTMLInputElement = canvas.getByRole('textbox', { name: /phone/i });
    const country: HTMLInputElement = canvas.getByRole('combobox', { name: /country/i });
    const street: HTMLInputElement = canvas.getByRole('textbox', { name: /street/i });
    const apartment: HTMLInputElement = canvas.getByRole('textbox', { name: /apartment/i });
    const city: HTMLInputElement = canvas.getByRole('textbox', { name: /city/i });
    const region: HTMLInputElement = canvas.getByRole('textbox', { name: /state/i });
    const zip: HTMLInputElement = canvas.getByRole('textbox', { name: /zip/i });
    const buttonCheckout: HTMLButtonElement = canvas.getByText('Checkout');
    expect(firstname.value).toBe(user.firstname);
    expect(lastname.value).toBe(user.lastname);
    expect(email.value).toBe(user.email);
    expect(phone.value).toBe(user.phone);
    expect(country.value).toBe(address.country);
    expect(street.value).toBe(address.street);
    expect(apartment.value).toBe(address.apartment);
    expect(city.value).toBe(address.city);
    expect(region.value).toBe(address.region);
    expect(zip.value).toBe(address.zip);
    expect(buttonCheckout).not.toBeDisabled();
  },
};

export const WithoutValidShippingInformation: Story = {
  decorators: [
    applicationConfig({
      providers: [{ provide: UserFacade, useValue: mockUserFacade(user) }],
    }),
  ],
  play: async ({ canvasElement, args }) => {
    const canvas = within(canvasElement);
    const buttonCheckout: HTMLButtonElement = canvas.getByText('Checkout');
    expect(buttonCheckout).toBeDisabled();
  },
};
