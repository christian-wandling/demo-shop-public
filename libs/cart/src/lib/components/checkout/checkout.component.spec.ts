import { ComponentFixture, TestBed } from '@angular/core/testing';
import { CheckoutComponent } from './checkout.component';
import { CartFacade } from '../../cart.facade';
import { signal, WritableSignal } from '@angular/core';
import { UserFacade } from '@demo-shop/user';
import { ReactiveFormsModule } from '@angular/forms';
import { By } from '@angular/platform-browser';
import { provideRouter, Router } from '@angular/router';
import { CartItemResponse, UserResponse } from '@demo-shop/api';
import { provideMockImageLoader } from '@demo-shop/shared';

describe('CheckoutComponent', () => {
  const mockUser: WritableSignal<UserResponse> = signal({
    id: 1,
    email: 'john@doe.com',
    firstname: 'John',
    lastname: 'Doe',
    phone: '123-456789',
    address: {
      street: 'street',
      apartment: 'apartment',
      city: 'city',
      region: 'region',
      zip: 'zip',
      country: 'country',
    },
  });

  const cartItems: WritableSignal<CartItemResponse[]> = signal([
    {
      id: 0,
      productId: 0,
      productName: 'productName',
      productThumbnail: 'productThumbnail',
      quantity: 0,
      unitPrice: 0,
      totalPrice: 0,
    },
  ]);

  let component: CheckoutComponent;
  let fixture: ComponentFixture<CheckoutComponent>;
  let cartFacade: CartFacade;
  let userFacade: UserFacade;
  let router: Router;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [CheckoutComponent, ReactiveFormsModule],
      providers: [
        provideMockImageLoader(),
        Router,
        {
          provide: CartFacade,
          useValue: {
            getAll: jest.fn().mockReturnValue(cartItems),
            getTotalPrice: jest.fn().mockReturnValue(signal(0)),
            removeItem: jest.fn().mockResolvedValue(undefined),
            loadShoppingSession: jest.fn().mockResolvedValue(undefined),
            checkout: jest.fn(),
          },
        },
        {
          provide: UserFacade,
          useValue: {
            getCurrentUser: jest.fn().mockReturnValue(mockUser),
            updateUserAddress: jest.fn(),
            updateUserPhone: jest.fn(),
          },
        },
        provideRouter([]),
      ],
    }).compileComponents();

    cartFacade = TestBed.inject(CartFacade);
    userFacade = TestBed.inject(UserFacade);
    router = TestBed.inject(Router);
    fixture = TestBed.createComponent(CheckoutComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match the snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });

  describe('checkout form', () => {
    it('should fill the form with user data and display it', () => {
      expect(component.checkoutForm().value).toEqual({
        phone: mockUser().phone,
        address: mockUser().address,
      });
      expect(fixture.debugElement.query(By.css('form'))).toBeTruthy();
    });
  });

  describe('update', () => {
    it('should enable update if form is dirty and valid', () => {
      component.checkoutForm().markAsDirty();
      fixture.detectChanges();

      const res = component.getUpdateEnabled();

      expect(res()).toEqual(true);
    });

    it('should disable update if form is not dirty', () => {
      component.checkoutForm().markAsPristine();
      fixture.detectChanges();

      const res = component.getUpdateEnabled();

      expect(res()).toEqual(false);
    });
  });

  describe('checkout', () => {
    it('should enable checkout if all conditions are filled', () => {
      const res = component.getCheckoutEnabled();

      expect(res()).toEqual(true);
    });

    it('should disable checkout if cart is empty', () => {
      cartItems.set([]);
      fixture.detectChanges();

      const res = component.getCheckoutEnabled();

      expect(res()).toEqual(false);
    });

    it('should disable checkout if user address is empty', () => {
      mockUser.set({ ...mockUser(), address: undefined });

      fixture.detectChanges();

      const res = component.getCheckoutEnabled();

      expect(res()).toEqual(false);
    });

    it('should disable checkout if form is dirty', () => {
      component.checkoutForm().markAsDirty();
      fixture.detectChanges();

      const res = component.getCheckoutEnabled();

      expect(res()).toEqual(false);
    });

    it('should start the checkout process', async () => {
      jest.spyOn(router, 'navigateByUrl');

      await component.checkout();

      expect(cartFacade.checkout).toHaveBeenCalled();
      expect(router.navigateByUrl).toHaveBeenCalledWith('/products');
    });
  });

  describe('cart-items', () => {
    it('should display the cart items', () => {
      expect(fixture.debugElement.query(By.css('lib-cart-items'))).toBeTruthy();
    });

    it('should remove an item', () => {
      const id = 1;

      component.removeItem(id);

      expect(cartFacade.removeItem).toHaveBeenCalledWith(id);
    });
  });

  describe('update user', () => {
    it('should update the phone if changed', async () => {
      mockUser.set({ ...mockUser(), phone: '123' });
      const ctrl = component.checkoutForm().controls.phone;
      ctrl?.markAsDirty();

      await component.updateUser();

      expect(userFacade.updateUserPhone).toHaveBeenCalledWith({ phone: ctrl?.value });
    });

    it('should update the address if changed', async () => {
      mockUser.set({
        ...mockUser(),
        address: {
          street: 'street',
          apartment: 'apartment',
          city: 'city',
          region: 'region',
          zip: 'zip',
          country: 'country',
        },
      });
      const ctrl = component.checkoutForm().controls.address;
      ctrl?.markAsDirty();

      await component.updateUser();

      expect(userFacade.updateUserAddress).toHaveBeenCalledWith(ctrl?.value);
    });
  });
});
