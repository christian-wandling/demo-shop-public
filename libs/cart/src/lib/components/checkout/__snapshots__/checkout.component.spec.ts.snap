// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckoutComponent should match the snapshot 1`] = `
<lib-checkout
  BREAKPOINT_MD={[Function Number]}
  checkoutForm={[Function Function]}
  items={[Function Function]}
  price={[Function Function]}
  shippingInformationExtended={[Function Function]}
  user={[Function Function]}
>
  <div
    class="bg-gray-100 sm:m-2 lg:m-6 p-4 sm:p-8 lg:p-10 sm:rounded"
  >
    <div
      class="flex flex-col lg:flex-row justify-between gap-8"
    >
      <form
        class="ng-untouched ng-pristine ng-valid"
        novalidate=""
      >
        <div
          class="space-y-8"
        >
          <div>
            <div
              class="flex justify-between items-center p-4 bg-gray-100 cursor-pointer lg:pointer-events-none"
            >
              <h2
                class="text-base font-semibold leading-7 text-gray-900"
              >
                Shipping information
              </h2>
              <svg
                class="lg:hidden w-5 h-5 transition-transform duration-300 rotate-180"
                fill="none"
                stroke="currentColor"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19 9l-7 7-7-7"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
            </div>
            <div
              class="mt-4 lg:max-h-[1000px] grid md:grid-cols-8 gap-x-6 gap-y-4 transition-all duration-500 ease-in-out overflow-hidden max-h-[1000px]"
            >
              <div
                class="md:col-span-4"
              >
                <label
                  class="required block text-sm font-medium leading-6 text-gray-500 disabled"
                  for="first-name"
                >
                   First name 
                </label>
                <div
                  class="mt-2"
                >
                  <input
                    autocomplete="given-name"
                    class="block w-full rounded-md border-0 py-1.5 text-gray-500 bg-gray-200 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6 ng-untouched ng-pristine"
                    disabled=""
                    formcontrolname="firstname"
                    id="first-name"
                    name="first-name"
                    type="text"
                  />
                </div>
                <lib-form-error
                  class="text-xs text-red-600"
                >
                </lib-form-error>
              </div>
              <div
                class="md:col-span-4"
              >
                <label
                  class="required block text-sm font-medium leading-6 text-gray-500"
                  for="last-name"
                >
                   Last name 
                </label>
                <div
                  class="mt-2"
                >
                  <input
                    autocomplete="family-name"
                    class="block w-full rounded-md border-0 py-1.5 text-gray-500 bg-gray-200 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6 ng-untouched ng-pristine"
                    disabled=""
                    formcontrolname="lastname"
                    id="last-name"
                    name="last-name"
                    type="text"
                  />
                  <lib-form-error
                    class="text-xs text-red-600"
                  >
                  </lib-form-error>
                </div>
              </div>
              <div
                class="md:col-span-4"
              >
                <label
                  class="required block text-sm font-medium leading-6 text-gray-500"
                  for="email"
                >
                   Email address 
                </label>
                <div
                  class="mt-2"
                >
                  <input
                    autocomplete="email"
                    class="block w-full rounded-md border-0 py-1.5 text-gray-500 bg-gray-200 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6 ng-untouched ng-pristine"
                    disabled=""
                    formcontrolname="email"
                    id="email"
                    name="email"
                    type="email"
                  />
                </div>
                <lib-form-error
                  class="text-xs text-red-600"
                >
                </lib-form-error>
              </div>
              <div
                class="md:col-span-4"
              >
                <label
                  class="required block text-sm font-medium leading-6 text-gray-900"
                  for="phone"
                >
                  Phone
                </label>
                <div
                  class="mt-2"
                >
                  <input
                    autocomplete="tel"
                    class="block w-full rounded-md border-0 py-1.5 text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6 ng-untouched ng-pristine ng-valid"
                    formcontrolname="phone"
                    id="phone"
                    name="phone"
                  />
                </div>
              </div>
              <div
                class="md:col-span-4"
              >
                <label
                  class="required block text-sm font-medium leading-6 text-gray-900"
                  for="country"
                >
                  Country
                </label>
                <div
                  class="mt-2"
                >
                  <select
                    autocomplete="country-name"
                    class="block w-full rounded-md border-0 py-1.5 text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:max-w-xs sm:text-sm sm:leading-6 ng-untouched ng-pristine ng-valid"
                    formcontrolname="country"
                    id="country"
                    name="country"
                  >
                    <option>
                      United States
                    </option>
                    <option>
                      Austria
                    </option>
                    <option>
                      Vietnam
                    </option>
                  </select>
                </div>
                <lib-form-error
                  class="text-xs text-red-600"
                >
                </lib-form-error>
              </div>
              <div
                class="md:col-span-5"
              >
                <label
                  class="required block text-sm font-medium leading-6 text-gray-900"
                  for="street"
                >
                  Street
                </label>
                <div
                  class="mt-2"
                >
                  <input
                    autocomplete="address-line1"
                    class="block w-full rounded-md border-0 py-1.5 text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6 ng-untouched ng-pristine ng-valid"
                    formcontrolname="street"
                    id="street"
                    name="street"
                    type="text"
                  />
                </div>
              </div>
              <div
                class="md:col-span-3"
              >
                <label
                  class="required block text-sm font-medium leading-6 text-gray-900"
                  for="apartment"
                >
                   Apartment 
                </label>
                <div
                  class="mt-2"
                >
                  <input
                    autocomplete="address-line2"
                    class="block w-full rounded-md border-0 py-1.5 text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6 ng-untouched ng-pristine ng-valid"
                    formcontrolname="apartment"
                    id="apartment"
                    name="apartment"
                    type="text"
                  />
                </div>
                <lib-form-error
                  class="text-xs text-red-600"
                >
                </lib-form-error>
              </div>
              <div
                class="md:col-span-3"
              >
                <label
                  class="required block text-sm font-medium leading-6 text-gray-900"
                  for="city"
                >
                  City
                </label>
                <div
                  class="mt-2"
                >
                  <input
                    autocomplete="address-level2"
                    class="block w-full rounded-md border-0 py-1.5 text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6 ng-untouched ng-pristine ng-valid"
                    formcontrolname="city"
                    id="city"
                    name="city"
                    type="text"
                  />
                </div>
                <lib-form-error
                  class="text-xs text-red-600"
                >
                </lib-form-error>
              </div>
              <div
                class="md:col-span-3"
              >
                <label
                  class="block text-sm font-medium leading-6 text-gray-900"
                  for="region"
                >
                  State / Province
                </label>
                <div
                  class="mt-2"
                >
                  <input
                    autocomplete="address-level1"
                    class="block w-full rounded-md border-0 py-1.5 text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6 ng-untouched ng-pristine ng-valid"
                    formcontrolname="region"
                    id="region"
                    name="region"
                    type="text"
                  />
                </div>
              </div>
              <div
                class="md:col-span-2"
              >
                <label
                  class="required block text-sm font-medium leading-6 text-gray-900"
                  for="zip"
                >
                   ZIP / Postal code 
                </label>
                <div
                  class="mt-2"
                >
                  <input
                    autocomplete="postal-code"
                    class="block w-full rounded-md border-0 py-1.5 text-gray-900 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6 ng-untouched ng-pristine ng-valid"
                    formcontrolname="zip"
                    id="zip"
                    name="zip"
                    type="text"
                  />
                </div>
                <lib-form-error
                  class="text-xs text-red-600"
                >
                </lib-form-error>
              </div>
              <button
                class="btn-update md:col-span-full float-right mt-6 lg:max-w-60 flex w-full items-center justify-center rounded-md border border-transparent px-8 py-3 text-base font-medium bg-gray-200 text-gray-800 pointer-events-none"
                disabled=""
              >
                 Update 
              </button>
            </div>
          </div>
        </div>
      </form>
      <section>
        <h2
          class="text-lg font-medium text-gray-900"
          id="slide-over-title"
        >
          Order Summary
        </h2>
        <lib-cart-items>
          <div
            class="mt-8"
          >
            <div
              class="flow-root"
            >
              <ul
                class="-my-6 divide-y divide-gray-200"
              >
                <li
                  class="flex py-6"
                >
                  <div
                    class="relative h-24 w-24 aspect-1 flex-shrink-0 overflow-hidden rounded-md border border-gray-200 group-hover:opacity-75 cursor-pointer"
                    tabindex="0"
                  >
                    <img
                      alt="productName"
                      fetchpriority="auto"
                      fill=""
                      loading="lazy"
                      ng-img="true"
                      placeholder=""
                      sizes="100vw"
                      src="productThumbnail"
                      srcset="productThumbnail 640w, productThumbnail 750w, productThumbnail 828w, productThumbnail 1080w, productThumbnail 1200w, productThumbnail 1920w, productThumbnail 2048w, productThumbnail 3840w"
                      style="position: absolute; width: 100%; height: 100%; inset: 0; background-size: cover; background-position: 50% 50%; background-repeat: no-repeat; background-image: url(productThumbnail); filter: blur(15px);"
                      title="productName"
                    />
                  </div>
                  <div
                    class="ml-4 flex flex-1 flex-col"
                  >
                    <div>
                      <div
                        class="flex justify-between text-base font-medium text-gray-900"
                      >
                        <h3>
                          <a>
                            productName
                          </a>
                        </h3>
                        <p
                          class="ml-4"
                        >
                          $0.00
                        </p>
                      </div>
                    </div>
                    <div
                      class="flex flex-1 items-end justify-between text-sm"
                    >
                      <p
                        class="text-gray-500"
                      >
                        Qty 0
                      </p>
                      <div
                        class="flex"
                      >
                        <button
                          class="font-medium text-indigo-600 hover:text-indigo-500"
                          type="button"
                        >
                           Remove 
                        </button>
                      </div>
                    </div>
                  </div>
                </li>
              </ul>
            </div>
          </div>
        </lib-cart-items>
        <div
          class="border-t border-gray-900/10 mt-12"
        >
          <div
            class="mt-8 flex justify-between"
          >
            <span>
              Total
            </span>
            <span>
              $0.00
            </span>
          </div>
        </div>
        <section>
          <div
            class="mt-6 flex items-center justify-center gap-x-6"
          >
            <button
              class="btn-checkout mt-6 lg:max-w-60 flex w-full items-center justify-center rounded-md border border-transparent px-8 py-3 text-base font-medium bg-indigo-600 text-white hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-indigo-500 focus:ring-offset-2"
            >
               Checkout 
            </button>
          </div>
        </section>
      </section>
    </div>
  </div>
</lib-checkout>
`;
